### PR TITLE
D pattison/state estimator

### PR DESCRIPTION
**Pull request by dpattison3**
Still need to collect GPS survey of origin on track for the launch parameter and update the start GPS coordinates in Gazebo accordingly.

### CODE REVIEW

File: autorally_core/src/StateEstimator/BlockingQueue.h [hunk 1]
```diff
@@ -90,6 +90,18 @@ class BlockingQueue {
     return boost::optional<WorkType>(element);
   }
 
+  // TODO this might not make sense - will assess later
```
Comment (pdrews) [inline autorally_core/src/StateEstimator/BlockingQueue.h]: Why did you add this?  It doesn't seem to be used, if not lets remove it from the PR

File: autorally_core/src/StateEstimator/StateEstimator.cpp [hunk 2]
```diff
@@ -284,42 +284,113 @@ namespace autorally_core
...
-      ++m_gpsCounter;
-      if (m_gpsCounter > m_gpsSkip) m_gpsCounter = 0;
-      else continue;
+      sensor_msgs::NavSatFixConstPtr fix;
+      bool usingGPS = false;
+      bool usingOdom = false;
+      bool usingIMU = false;
+
+
+      // remove any old GPS measurements
```
Comment (pdrews) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: Is it possible this will cause us to throw out gps measurements that have a place in the graph?

File: autorally_core/src/StateEstimator/BlockingQueue.h (hunk 1 above)
Comment (pdrews) [inline autorally_core/src/StateEstimator/BlockingQueue.h]: I was wrong, it is used.  This breaks the initial design goal of the queue class, but maybe that is ok in our case

File: autorally_core/src/StateEstimator/StateEstimator.cpp [hunk 3]
```diff
@@ -284,42 +284,113 @@ namespace autorally_core
...
+      // set up for using odom or GPS timestamp but always at ~10Hz
+      if (m_gotFirstFix && m_gpsOptQ.size() > 0 && (m_gpsOptQ.back()->header.stamp.toSec() - prevTime) > 0.9/m_frequency)
+      {
+        fix = m_gpsOptQ.popBlocking();
+        while ((fix->header.stamp.toSec() - prevTime) < 0.9/m_frequency)
+          fix = m_gpsOptQ.popBlocking();
+
+        // check if the GPS measurement is close to expected from the previous velocity and pose
+        double E, N, U, expectedE, expectedN, expectedU, distance;
+        m_enu.Forward(fix->latitude, fix->longitude, fix->altitude, E, N, U);
```
Comment (pdrews) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: We already push the pose forward using the IMU, why don't we use that estimate here instead of pretending we are going at a constant velocity?

File: autorally_core/src/StateEstimator/StateEstimator.cpp [hunk 4]
```diff
@@ -284,42 +284,113 @@ namespace autorally_core
...
+        else
+        {
+          ROS_WARN("GPS message does not match expected position");
+          diag("GPS error count", std::to_string(++gpsErrorCount));
+        }
+      }
+      else if (m_usingOdom && m_gotFirstFix && (m_odomOptQ.size() > 0)
+          && (m_odomOptQ.back()->header.stamp.toSec() - prevTime) > 1.25/m_frequency)
+      {
+        // GPS messages are a little slower so wait a little extra before continuing without one
```
Comment (pdrews) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: This seems pretty hacky.  We don't want to throw away a gps message just because it was a little too late.  We can put the odometry in the graph and come back and add the gps measurement later

File: autorally_core/src/StateEstimator/StateEstimator.cpp [hunk 5]
```diff
@@ -345,125 +416,178 @@ namespace autorally_core
...
-        m_prevPose = m_isam->calculateEstimate<Pose3>(X(m_poseVelKey+1));
-        m_prevVel = m_isam->calculateEstimate<Vector3>(V(m_poseVelKey+1));
-        m_previousBias = m_isam->calculateEstimate<imuBias::ConstantBias>(B(m_biasKey+1));
-        //std::cout << m_isam->marginalCovariance(X(m_poseVelKey+1)) << std::endl << std::endl;
-
-        diag_ok("Still ok!");
 
+        try
         {
-          boost::mutex::scoped_lock guard(m_optimizedStateMutex);
```
Comment (pdrews) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: Why did the call to isam->update move outside of the lock??  We add factors to the graph in another thread, and I don't think calls into gtsam are threadsafe.  This mutex should be protecting both.

Comment (dpattison3) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: In the original, update isn't in the thread guard. I didn't remove anything from it.

File: autorally_core/src/StateEstimator/StateEstimator.cpp (hunk 4 above)
Comment (dpattison3) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: Not a bad idea. How would we know which position key each GPS message would go to?

Comment (dpattison3) [inline autorally_core/src/StateEstimator/StateEstimator.cpp]: Also, how would we account for the GPS timestamp being different than the timestamps used for the IMU (possibly with odom) factor? Without the GPS measurement there already, we wouldn't know what its timestamp will be. Therefore, we wouldn't necessarily have a position variable to add the GPS measurement to.
